src/libmongoc/src/mongoc/mongoc-client-side-encryption.c

```diff
@@ -1694,6 +1788,183 @@ mongoc_client_encryption_create_datakey (
+      }
+
+      BSON_ASSERT (bson_iter_recurse (&iter, &iter));
+
+      while (bson_iter_next (&iter)) {
+         const uint8_t *data = NULL;
+         uint32_t len = 0u;
+         bson_t key;
+         bson_iter_t key_iter;
+         bson_subtype_t subtype;
+         bson_t selector = BSON_INITIALIZER;
+         bson_t document = BSON_INITIALIZER;
+
+         bson_iter_document (&iter, &len, &data);
+
+         if (!data || !bson_init_static (&key, data, len)) {
+            bson_set_error (error,
+                            MONGOC_ERROR_CLIENT,
+                            MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_STATE,
+                            "element is not a valid BSON document");
+            GOTO (fail);
+         }
+
+         /* Find _id and use as selector. */
+         {
+            if (!bson_iter_init_find (&key_iter, &key, "_id")) {
+               bson_set_error (error,
+                               MONGOC_ERROR_CLIENT,
+                               MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_STATE,
+                               "could not find _id in key document");
+               GOTO (fail);
+            }
+
+            bson_iter_binary (&key_iter, &subtype, &len, &data);
+
+            if (!data || subtype != BSON_SUBTYPE_UUID) {
+               bson_set_error (error,
+                               MONGOC_ERROR_CLIENT,
+                               MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_STATE,
+                               "expected _id in key document to be a UUID");
+               GOTO (fail);
+            }
+
+            BSON_ASSERT (bson_append_iter (&selector, "_id", 3, &key_iter));
+         }
+
+         /* Find and include potentially updated fields. */
+         {
+            bson_t child;
+
+            BSON_ASSERT (
+               BSON_APPEND_DOCUMENT_BEGIN (&document, "$set", &child));
+            {
+               if (bson_iter_init_find (&key_iter, &key, "masterKey")) {
+                  BSON_ASSERT (
+                     bson_append_iter (&child, "masterKey", -1, &key_iter));
+               }
+
+               if (bson_iter_init_find (&key_iter, &key, "keyMaterial")) {
+                  BSON_ASSERT (
+                     bson_append_iter (&child, "keyMaterial", -1, &key_iter));
+               }
+            }
+            BSON_ASSERT (bson_append_document_end (&document, &child));
+         }
+
+         /* Update updateDate field. */
+         BCON_APPEND (&document,
+                      "$currentDate",
+                      "{",
+                      "updateDate",
+                      BCON_BOOL (true),
+                      "}");
+
+         if (!mongoc_bulk_operation_update_one_with_opts (
+                bulk, &selector, &document, NULL, error)) {
+            GOTO (fail);
+         }
+
+         bson_destroy (&key);
+         bson_destroy (&selector);
+      }
+
+      if (!mongoc_bulk_operation_execute (bulk, bulk_write_result, error)) {
+         GOTO (fail);
```

Comment:
I assume `bson_set_error` is not used here because you're exposing the underlying bulk write error, but do you care about potentially reporting a different error domain/code? Everything above seemed to use `MONGOC_ERROR_CLIENT` and `MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_STATE`, respectively.

Reply:
cc: @kevinAlbs

Reply:
SGTM. I'll resolve this.